src/openfermion/measurements/_bayesian_estimators.py

```diff
@@ -0,0 +1,845 @@
+
+    More details can be found in arXiv:1809.09697, Appendix C.
+    """
+
+    def __init__(self,
+                 amplitude_guess=None,
+                 amplitude_vars=None,
+                 num_vectors=1,
+                 num_freqs=1000,
+                 max_n=1,
+                 vector_guess=None):
+        """
+        Args:
+            amplitude_guess (numpy array or list):
+                estimates for the amplitude of the jth component of
+                the estimated function.
+            amplitude_vars (numpy array or list): variance on the
+                guess in amplitude_guess.
+            num_vectors (int): number of phases phi_j to estimate
+            num_freqs (int): number of wave components cos(n*phi_j)
+                to store coefficients for (this provides a cut-off to
+                the sensitivity of the function).
+            max_n (int): the maximum n to be used when multiplying
+                the stored function by matrices of the form
+                cos(n*phi_j + beta).
+            vector_guess (numpy array): a prior estimate of the function,
+                given in terms of the Fourier components c_{j,n}.
+                If None, estimated as a flat function over [-pi,pi].
+
+        Raises:
+            ValueError: if prior distribution does not have required shape.
+            ValueError: if insufficient amplitude guesses are given.
+            ValueError: if starting amplitudes do not break symmetry.
+        """
+
+        # Store size data
+        self._num_vectors = num_vectors
+        self._num_freqs = num_freqs
+
+        if amplitude_guess is None:
+            amplitude_guess = [1]
+        if amplitude_vars is None:
+            amplitude_vars = [[0]]
+
+        # Store prior information
+        if vector_guess is not None and (
+                num_vectors != vector_guess.shape[1]
+                or 2*num_freqs+1 != vector_guess.shape[0]):
+            raise ValueError('''Prior distribution of phases has shape {},
+                required shape is {}'''.format(
+                    vector_guess.shape, (2*num_freqs+1, num_vectors)))
+
+        self._vector_guess = vector_guess
+
+        if num_vectors != len(amplitude_guess):
+            raise ValueError('''I need {} amplitudes, but you have
+                given me {}.'''.format(num_vectors, len(amplitude_guess)))
+
+        self._amplitude_guess = amplitude_guess
+        self._amplitude_vars = amplitude_vars
+        self._amplitude_estimates = None
+        self._fourier_vectors = None
+
+        if any([numpy.isclose(self._amplitude_guess[j],
+                              self._amplitude_guess[j+1])
+                for j in range(num_vectors - 1)]):
+            raise ValueError('''Starting amplitudes must be different.''')
+
+        # Store max_n
+        self._max_n = max_n
+
+        # Create matrices for multiplication
+        self._make_matrices()
+        self.reset()
+
+    def reset(self):
```

Comment:
I thought we were getting rid of this method?

Reply:
Sorry, which method?

Reply:
`reset`

Reply:
Oh, I hadn't planned on doing so. The main reason why I want to have the reset option is to prevent having to remake _matrices every time the user does a new estimation, as this is quite costly. Which I guess is just heading towards the idea of 'separate the class and the functions which act on the class', but it doesn't feel completely right to have these functions inside the estimator itself...

Perhaps I should make a third class containing these matrices? Then the Bayesian estimator could work by creating a new FourierProbabilityDist every round, rather than copying and replacing all of the contents every single time; this might be a neater solution.